### ISSUE TITLE
cmd/govim: disruptive warning when editing non-Go files

### ISSUE THREAD

**Issue by cespare** [govim/govim#692]
After updating to latest master, I'm getting a warning dialog from govim when editing non-Go files.

For instance, here's what I see when I edit my .vimrc:

![screen_20200119122705](https://user-images.githubusercontent.com/318501/72687876-11c87f00-3ab7-11ea-9568-6e8a7e6aebc1.png)

The text says:

> You are neither in a module nor in your GOPATH. Please see https://github.com/golang/go/wiki/Modules for information on how to set up your Go project.

This seems similar to the issue discussed in https://github.com/golang/go/issues/35818.

A bisect shows that this change occurred in d80f0e21c2e3291f93bd60d3d70079f877834d85.

**Comment by myitcv** [comment]
For some reason I can't reproduce this:

```
cd $(mktemp -d)
touch blah.txt
vi blah.txt
```

It doesn't even reproduce when opening `.vimrc` from `$HOME` (although the initial `go list` does take forever, per https://github.com/golang/go/issues/35818)

I agree we need to do something about this with `govim` because it will annoy some people, but I think this likely needs to be balanced with having suitable messages for people who are just getting started.

Let's get this reproduced first then I think we look to some sort of "I know what I'm doing" config option.

To my mind https://github.com/golang/go/issues/35818 still needs to be fixed because it's possible to determine one is outside `GOPATH` and module context without running that expensive query.

**Comment by cespare** [comment]
Weird. It doesn't happen in a tempdir, or my homedir, or anywhere else that I've found *except* my `~/.vim` (git repo). Still working out what's different about this subtree.

> Let's get this reproduced first then I think we look to some sort of "I know what I'm doing" config option.

I suspect this is just some kind of bug. There's really nothing useful for anyone to get from this message -- I'm editing a non-Go file in a non-Go project; I don't want to see anything from govim.

**Comment by cespare** [comment]
OK, the problem is that the vim repo contains some Go code deep inside, and gopls can't figure out the dependencies.

Here's how to repro:

```
cd $(mktemp -d)
mkdir -p x/y
cat >x/y/z.go <<EOF
package main

import "example.com/blah"
_ = blah.Foo
func main() {}
EOF
vim ~/.zshrc # or some other unrelated file, or indeed any file
```

Even if there is some totally malformed Go project that gopls cannot understand in some subtree, I don't want govim to tell me about it while I'm editing my .vimrc.

**Comment by myitcv** [comment]
Thanks for the repro. 

I suspect, given your repro, that I would actually have seen the same error message had I left my Vim with a working directory of `$HOME` open long enough for the massive `go list` reported in golang/go#35818 to complete.

I think this confirms my current feelings:

1. we should be able to determine we are outside of `GOPATH` and module context without this expensive `go list`
2. we need a "I know what I'm doing" mode for `govim` 

Let's make this issue a tracker for point 2. We will, I suspect, need some option in `gopls` to support this fully so I'll raise a separate issue for that.

**Comment by cespare** [comment]
> 2\. we need a "I know what I'm doing" mode for `govim`

I feel like we're having trouble understanding each other. The misunderstanding is probably on my end; please explain where I've gone wrong.

Suppose that Alice installs govim. Alice doesn't have any particular govim expertise and she does not  select the "I know what I'm doing" mode. Alice uses govim for Go work; works great. Okay.

Now Alice runs `vim ~/.vimrc`. Is it your position that, at this point, vim (because govim is running, even though no Go is in play) ought to display a warning message?

**Comment by myitcv** [comment]
> I feel like we're having trouble understanding each other. The misunderstanding is probably on my end; please explain where I've gone wrong.

The misunderstanding was in fact entirely due to my not having thought through how point 2 in https://github.com/govim/govim/issues/692#issuecomment-577092935 would actually work 😄 

In any case, I've just merged bb606ba which I think will actually make my suggestion entirely moot. Because we now no longer have an expensive `go list` run at startup (https://github.com/golang/go/issues/35818) and it is my understanding that you will only see that error if you try to open a `.go` file outside a module or not within `GOPATH`, which seems fair.

That said, I've also seen reports (not tracked in an issue) that the warning message is not showing up when it should... so we might end up re-opening this at some point. I'll close for now because, the problem as reported here has gone (for now).
